src/tactic/rewrite_search/search.lean

```diff
@@ -0,0 +1,212 @@
+/-
+Copyright (c) 2020 Kevin Lacker, Keeley Hoek, Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Lacker, Keeley Hoek, Scott Morrison
+-/
+
+import meta.rb_map
+import tactic.rewrite_search.discovery
+import tactic.rewrite_search.types
+
+/-!
+# The graph algorithm part of rewrite search
+
+`search.lean` contains the logic to do a graph search. The search algorithm
+starts with an equation to prove, treats the left hand side and right hand side as
+two vertices in a graph, and uses rewrite rules to find a path that connects the two sides.
+-/
+
+open tactic
+
+namespace tactic.rewrite_search
+
+/--
+An edge represents a proof that can get from one expression to another.
+It represents the fact that, starting from the vertex `fr`, the expression in `proof`
+can prove the vertex `to`.
+`how` contains information that the explainer will use to generate Lean code for the
+proof.
+-/
+meta structure edge :=
+(fr to : ℕ)
+(proof : tactic expr)
+(how   : how)
+
+/-- Converting an edge to a human-readable string. -/
+meta def edge.to_string : edge → format
+| e := format!"{e.fr} → {e.to}"
+
+meta instance edge.has_to_format : has_to_format edge := ⟨edge.to_string⟩
+
+/--
+A vertex represents an expression that is equivalent to either the left or right side
+of our initial equation.
+`id` is a numerical id used to refer to this vertex in the context of a single graph.
+`exp` is the expression this vertex represents.
+`pp` is the string format of the expression; we store this in the vertex to avoid
+recalculating it.
+`side` is whether this vertex descends from the left or right side of the equation.
+`parent` is the edge that originally added this vertex to the graph.
```

Comment:
```suggestion
* `id` is a numerical id used to refer to this vertex in the context of a single graph.
* `exp` is the expression this vertex represents.
* `pp` is the string format of the expression; we store this in the vertex to avoid
  recalculating it.
* `side` is whether this vertex descends from the left or right side of the equation.
* `parent` is the edge that originally added this vertex to the graph.
```

Reply:
OK